Data for modeler (epic 4.1)

guilpier-code opened the pull request:
Associated to [epic 4.1](https://gopro-collaboratif.rte-france.com/pages/viewpage.action?pageId=411149942).
Associated to [ticket  ANT-2031](https://gopro-tickets.rte-france.com/browse/ANT-2031)

guilpier-code committed:
Data for new modeler : make class linearProblemData derive from an interface

guilpier-code committed:
Data for new modeler (4.1) : passing new interface ILinearProblemData instead of old concrete LinearProblemData

guilpier-code commented on src/solver/modeler/api/include/antares/solver/modeler/api/linearProblemData.h:
```diff
@@ -28,20 +28,33 @@
+                   unsigned int scenario,
+                   unsigned int hour) override;
+
+//    unsigned getTimeResolutionInMinutes();
+//    bool hasScalarData(const std::string& key);
+//    double getScalarData(const std::string& key, unsigned scenario);
+//    bool hasTimedData(const std::string& key);
+//    const std::vector<double>& getTimedData(const std::string& key, unsigned scenario);
+
+//private:
+//    std::vector<int> timeStamps_;
+//    unsigned timeResolutionInMinutes_;
+//    std::map<std::string, std::vector<double>> scalarData_;
+//    std::map<std::string, std::vector<std::vector<double>>> timedData_;
```
This data are now commented out.
It will be uncommented if we needed eventually.

guilpier-code commented on src/solver/modeler/api/include/antares/solver/modeler/api/linearProblemData.h:
```diff
@@ -28,20 +28,33 @@
 {
 
-class LinearProblemData
+class ILinearProblemData
```
Interface class **ILinearProblemData** was created here, but it should be eventually moved in a separate header file.
The current location (**solver/modeler/api**) seems the right place for this header file, because this interface belongs to and should be defined where the contract it represents is actually required : class **LinearProblemBuilder** and class **LinearProblemFiller**

guilpier-code commented on src/solver/modeler/api/include/antares/solver/modeler/api/linearProblemData.h:
```diff
@@ -28,20 +28,33 @@
+                           unsigned int hour) = 0;
+};
+
+class LinearProblemData : public ILinearProblemData
```
About where we should put class **LinearProblemData** (which is a implementation of interface **ILinearProblemData**) : 
We should take [this page](https://gopro-collaboratif.rte-france.com/pages/viewpage.action?pageId=424772626) (see part 3) into account.
Possible location candidates are : 
- **src/study/data-series**
- **src/data-series**

a-zakir replied:
With legacy code?
I’d rather create a specific implementation in that place.

guilpier-code replied:
We have a target of a new source file architecture [here](https://gopro-collaboratif.rte-france.com/pages/viewpage.action?pageId=424772626).
I guess handling new modeler data should go to one of sub directory suggested above, whether we mix legacy code or not. 
_To be decided_.

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : clang format

a-zakir commented:
It would be a good idea to write some tests that demonstrate the use cases of the new interface.

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : remove commented code

guilpier-code committed:
Data for new modeler (4.1) : adding a data series repository to class LinearProblemData

We plan to have one unique instance this latter class

guilpier-code committed:
Data for new modeler (4.1) : separate linear problem data implementation from its abstract parent

guilpier-code committed:
Data for new modeler (4.1) : packaging a data-series-lib library

guilpier-code committed:
Data for new modeler (4.1) : clang format

guilpier-code committed:
Data for new modeler (4.1) : coding scenario group repository

guilpier-code committed:
Data for new modeler (4.1) : testing scenario group repository

guilpier-code committed:
Data for new modeler (4.1) : avoid use of std::format(..)

guilpier-code committed:
Data for new modeler (4.1) : fix for Linux

guilpier-code committed:
Data for new modeler (4.1) : adding an implementation of IDataSeries : TimeSeriesSet

guilpier-code committed:
Data for new modeler (4.1) : coding the data series repository

guilpier-code committed:
Data for new modeler (4.1) : compilation fix

guilpier-code committed:
Data for new modeler (4.1) : clang format

guilpier-code committed:
Data for new modeler (4.1) : unit testing for time series set

guilpier-code committed:
Data for new modeler (4.1) : clang format

guilpier-code committed:
Data for new modeler (4.1) : mainly testind data series repository

guilpier-code committed:
Data for new modeler (4.1) : start for testing linear problem data

guilpier-code committed:
Data for new modeler (4.1) : moving linear problem data (from api to dataSeries) + cleaning

guilpier-code committed:
Data for new modeler (4.1) : tiny corrections

guilpier-code commented on src/solver/modeler/api/CMakeLists.txt:
```diff
@@ -8,13 +8,12 @@ set(SRC_API
         include/antares/solver/modeler/api/linearProblemFiller.h
         include/antares/solver/modeler/api/linearProblemBuilder.h
 
-        linearProblemData.cpp
```
class **LinearProblemData** was moved from here (**src/solver/modeler/api**) to **src/solver/modeler/dataSeries**

guilpier-code commented on src/solver/modeler/api/include/antares/solver/modeler/api/ILinearProblemData.h:
```diff
@@ -0,0 +1,18 @@
+namespace Antares::Solver::Modeler::Api
+{
+
+class ILinearProblemData
```
Concrete class **LinearProblemData** now inherits from abstract class **ILinearProblemData**

guilpier-code commented on src/tests/src/solver/modeler/api/CMakeLists.txt:
```diff
@@ -1,12 +1,11 @@
-  testModelerLinearProblemWithOrtools.cpp
-  testModelerLPbuilder.cpp
-  INCLUDE
-  "${src_solver_optimisation}"
```
**src_solver_optimisation** does not seem to be useful, we remove it

guilpier-code commented on src/tests/src/solver/optim-model-filler/CMakeLists.txt:
```diff
@@ -1,16 +1,15 @@
-  test_readLinearExpressionVisitor.cpp
-  test_readLinearConstraintVisitor.cpp
-  INCLUDE
-  "${src_solver_optimisation}"
```
**src_solver_optimisation** does not seem to be useful, we remove it

guilpier-code commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/dataSeriesRepo.h:
```diff
@@ -0,0 +1,24 @@
+class DataSeriesRepository
+{
+public:
+    void addDataSeries(std::unique_ptr<IDataSeries> dataSeries);
```
About `void addDataSeries(...)` : forces to **std::move** a unique pointer defined outside. 
So the client **IDataSeries** gets empty after this call.
I thought it's suitable, as **data series** have to end up inside the **data series repository** and seem useless somewhere else.

guilpier-code commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/timeSeriesSet.h:
```diff
@@ -0,0 +1,21 @@
+
+namespace Antares::Solver::Modeler::DataSeries
+{
+class TimeSeriesSet: public IDataSeries
```
Class **TimeSeriesSet** is supposed to be out of **epic 4.1**'s scope.
But we needed a concrete child of **IDataSeries** in order to unit test other classes (for instance **DataSeriesRepository**).
This class can end up not required and removed at some point. 
It comes with unit tests (see **testTimeSeriesSet.cpp**).

guilpier-code committed:
Data for new modeler (4.1) : adding unit tests on linear problem data

guilpier-code committed:
Data for new modeler (4.1) : adding unit tests on time series set

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/data-for-new-modeler--4.1

guilpier-code commented on src/solver/modeler/main.cpp:
```diff
@@ -23,21 +23,14 @@
 #include <antares/solver/modeler/parameters/parseModelerParameters.h>
 #include <antares/solver/optim-model-filler/ComponentFiller.h>
 
-#include "../optimisation/include/antares/solver/optimisation/LegacyFiller.h"
```
We take advantage of resolving conflicts to remove unnecessary code

flomnes committed:
Formatting

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/dataSeries.h:
```diff
@@ -0,0 +1,28 @@
+{
+public:
+    IDataSeries(std::string name):
+        name_(name)
```
```suggestion
    IDataSeries(std::string name):
        name_(std::move(name))
```

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/dataSeries.h:
```diff
@@ -0,0 +1,28 @@
+
+    virtual double getData(unsigned int rank, unsigned int hour) = 0;
+
+    std::string name()
+    {
+        return name_;
+    }
```
```suggestion
    std::string name() const
    {
        return name_;
    }
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/dataSeriesRepo.h:
```diff
@@ -0,0 +1,24 @@
+
+private:
+    std::map<std::string, std::unique_ptr<IDataSeries>> dataSeries_;
+    std::string err_prefix = "Data series repo : ";
```
Not useful

guilpier-code replied:
What is useless ? Could you be more specific ?
If you mean all of the private data members, please explain why and how we could do without them.

flomnes replied:
`err_prefix` becomes useless if we use custom exceptions

==> done

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/dataSeriesRepo.h:
```diff
@@ -0,0 +1,24 @@
+class DataSeriesRepository
+{
+public:
+    void addDataSeries(std::unique_ptr<IDataSeries> dataSeries);
```
```suggestion
    void addDataSeries(IDataSeries& dataSeries);
```

guilpier-code replied:
Please see [this answer](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2577#issuecomment-2607130421)

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/linearProblemData.h:
```diff
@@ -21,27 +21,30 @@
+    double getData(const std::string dataSetId,
+                   const std::string scenarioGroup,
+                   const unsigned scenario,
+                   const unsigned hour) override;
```
```suggestion
    double getData(const std::string& dataSetId,
                   const std::string& scenarioGroup,
                   const unsigned scenario,
                   const unsigned hour) override;
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/linearProblemData.h:
```diff
@@ -21,27 +21,30 @@
+                   const unsigned scenario,
+                   const unsigned hour) override;
+
+    void addScenarioGroup(std::string groupId, std::pair<unsigned, unsigned> scenarioToRank);
+    void addDataSeries(std::unique_ptr<IDataSeries> dataSeries);
```
```suggestion
    void addScenarioGroup(const std::string& groupId, std::pair<unsigned, unsigned> scenarioToRank);
    void addDataSeries(IDataSeries& dataSeries);
```

guilpier-code replied:
Please see [this answer](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2577#issuecomment-2607130421)

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/scenarioGroupRepo.h:
```diff
@@ -0,0 +1,19 @@
+    void addPairScenarioRankToGroup(std::string groupId,
+                                    std::pair<unsigned, unsigned> scenarioToRank);
+
+    unsigned getDataRank(std::string groupId, unsigned scenario);
```
```suggestion
    unsigned getDataRank(const std::string& groupId, unsigned scenario);
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/scenarioGroupRepo.h:
```diff
@@ -0,0 +1,19 @@
+{
+public:
+    void addPairScenarioRankToGroup(std::string groupId,
+                                    std::pair<unsigned, unsigned> scenarioToRank);
```
```suggestion
    void addPairScenarioRankToGroup(const std::string& groupId,
                                    std::pair<unsigned, unsigned> scenarioToRank);
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/timeSeriesSet.h:
```diff
@@ -0,0 +1,21 @@
+private:
+    unsigned height_ = 0;
+    std::vector<std::vector<double>> tsSet_;
+    std::string err_prefix_;
```
useless

guilpier-code replied:
What is useless ? Could you be more specific ?
If you mean all of the private data members, please explain why and how we could do without them.

flomnes replied:
The `err_prefix_` is not needed if you use custom exceptions

flomnes commented on src/solver/modeler/dataSeries/scenarioGroupRepo.cpp:
```diff
@@ -0,0 +1,31 @@
+    if (!scenarioGroups_.contains(groupId))
+    {
+        std::string error_message = "Group '" + groupId + "' does not exist in group repo.";
+        throw std::invalid_argument(error_message);
```
Use custom exceptions
```cpp
throw GroupDoesNotExist(groupId);
```

flomnes commented on src/solver/modeler/dataSeries/scenarioGroupRepo.cpp:
```diff
@@ -0,0 +1,31 @@
+    {
+        std::string error_message = "In scenario group '" + groupId + "', scenario '"
+                                    + std::to_string(scenario) + "' does not exist.";
+        throw std::invalid_argument(error_message);
```
```suggestion
        throw ScenarioDoesExist(groupId, scenario);
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/scenarioGroupRepo.cpp:
```diff
@@ -0,0 +1,31 @@
+  std::string groupId,
+  std::pair<unsigned, unsigned> scenarioToRank)
+{
+    scenarioGroups_[groupId] = {scenarioToRank};
```
If a key for `groupId` already exists, it will be erased, right ? Is this intended ?

guilpier-code replied:
No, it's not intended, I can also raise a (custom) exception in case we try to erase a scenario group with another.

guilpier-code replied:
Done

flomnes reviewed (commented):
- Use custom exceptions
- No need to pass `std::unique_ptr` ?

guilpier-code commented:
> * No need to pass `std::unique_ptr` ?

As said [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2577#discussion_r1923624038), passing **std::unique_ptr** is a way to force a piece of data to be held in one place only.
Once passed to another container with an **add()** method, the passed data gets empty outside the container, and the life time of the data becomes the life time of the container.
For example the **data series repository** is the only place to hold a **IDataSeries** (which can be a set of TS) because these **IDataSeries** were passed as **std::unique_ptr** .
Instead of storing a **std::unique_ptr** to a **IDataSeries** in the **data series repo**,  we could point it with a reference or a classic pointer, but what's the point of that ?
Are we supposed to reuse external TS independently once they are stored in the **LinearProblemData** ?

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review : introducing custom exceptions

flomnes commented on src/solver/modeler/dataSeries/include/antares/solver/modeler/dataSeries/timeSeriesSet.h:
```diff
@@ -0,0 +1,20 @@
+{
+public:
+    explicit TimeSeriesSet(std::string name, unsigned height);
+    void add(std::vector<double> ts);
```
```suggestion
    void add(const std::vector<double>& ts);
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/linearProblemData.cpp:
```diff
@@ -0,0 +1,47 @@
+namespace Antares::Solver::Modeler::DataSeries
+{
+
+void LinearProblemData::addScenarioGroup(std::string groupId,
```
```suggestion
void LinearProblemData::addScenarioGroup(const std::string& groupId,
```

guilpier-code replied:
Done

flomnes commented on src/solver/modeler/dataSeries/timeSeriesSetExceptions.cpp:
Use custom exceptions in this file

guilpier-code replied:
Done

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : correction due to review > add custom exceptions

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : adding an exception + unit test on scenario group repository

guilpier-code committed:
[skip ci] Data for new modeler (4.1) : add custom exceptions + add one check in production code

guilpier-code committed:
Data for new modeler (4.1) : more fix due to review

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into feature/data-for-new-modeler--4.1

flomnes committed:
Define exception classes as subclasses (#2612)

In order to avoid prefixes, we use sub-classes instead of separate
classes

For each class, we define a few sub-classes so that we don't need
prefixes.

flomnes committed:
Merge branch 'develop' into feature/data-for-new-modeler--4.1
